Attempt at getting a cc_library to select sdk_config.h based on binary

Michaelhobo opened the pull request:
So far, I can switch the sdk_config with the //:sdk_config=//a:sdk_config flag. But, I need to tie the setting to a cc_binary.

Michaelhobo committed:
Attempt at getting a cc_library to select sdk_config.h based on binary

So far, I can switch the sdk_config with the //:sdk_config=//a:sdk_config flag. But, I need to tie the setting based on the rule.

Michaelhobo commented:
it looks like this right now:

```bash
$ bazel run //a:bin --//:sdk_config=//a:sdk_config
INFO: Analyzed target //a:bin (0 packages loaded, 6 targets configured).
INFO: Found 1 target...
Target //a:bin up-to-date:
  bazel-bin/a/bin
INFO: Elapsed time: 0.309s, Critical Path: 0.17s
INFO: 1 process: 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
INFO: Build completed successfully, 2 total actions
a
```

```bash
$ bazel run //a:bin
INFO: Build option --//:sdk_config has changed, discarding analysis cache.
INFO: Analyzed target //a:bin (0 packages loaded, 57 targets configured).
INFO: Found 1 target...
Target //a:bin up-to-date:
  bazel-bin/a/bin
INFO: Elapsed time: 0.424s, Critical Path: 0.22s
INFO: 2 processes: 2 linux-sandbox.
INFO: Build completed successfully, 3 total actions
INFO: Build completed successfully, 3 total actions
default
```

bsilver8192 commented on BUILD:
```diff
@@ -0,0 +1,33 @@
+cc_library(
+    name = "lib",
+    hdrs = ["lib.h"],
+    includes = ["a"], # Why does replacing "a" with ":sdk_config_dir not work?
```
`string_flag` lets you access the result as a `BuildSettingInfo`. `includes` is looking for a list of strings, not a list of labels to extract strings from via their providers.

I would write a little custom rule (or genrule) which copies the file pointed to by `:sdk_config` into a file in a known place (like `sdk_config_dir/sdk_config.h`), and then you can pass `includes = ["sdk_config_dir"].

Some rules (not the C++ ones) have versions of their attrs which take a label, where they use the value of that file instead of a direct string ([pkg_deb](https://github.com/bazelbuild/rules_pkg/tree/master/pkg#pkg_deb).version_file for example). That allows solving this problem with a trivial custom rule/genrule that dumps the string into a file. Unfortunately, that doesn't help here.

Michaelhobo replied:
Would that cause problems if I build 2 binaries with conflicting sdk_config.h files?
bazel build //a:bin //b:bin

bsilver8192 replied:
Yea, you'd have to make a unique path for each one. You need a custom rule to do the transition anyways, so you could put it there, or a separate little rule. `ctx.label.name` is unique by definition, and `ctx.actions.declare_file` only lets you do stuff in the current package, so something like `ctx.actions.declare_file('%s_include_dir/sdk_config.h' % ctx.label.name)` works well.

bsilver8192 reviewed (commented):
I do know the answer to this question :)

Michaelhobo committed:
trying the transition

Michaelhobo commented:
I tried getting the do_transition from our conversation set up, but building //a:configured_bin is throwing an error I don't really understand. It seems like it's complaining that do_transition is executable (is this true? bazel run //a:configured_bin says no), and dep="//a:bin" is executable, but "//a:bin" isn't owned by //a:configured_bin? 

```bash
$ bazel build //a:configured_bin
ERROR: /home/michaelhobo/repos/bazel_playground/a/BUILD:10:14: in do_transition rule //a:configured_bin: 


<builtin>: 'executable' provided by an executable rule 'do_transition' should be created by the same rule.
ERROR: Analysis of target '//a:configured_bin' failed; build aborted: Analysis of target '//a:configured_bin' failed
INFO: Elapsed time: 0.117s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
```

Michaelhobo commented:
> I tried getting the do_transition from our conversation set up, but building //a:configured_bin is throwing an error I don't really understand. It seems like it's complaining that do_transition is executable (is this true? bazel run //a:configured_bin says no), and dep="//a:bin" is executable, but "//a:bin" isn't owned by //a:configured_bin?

I was under the impression that just depending on //a:bin from //a:configured_bin would create //a:bin - so I'm not really sure what's up with this. Do rules have to do something with their deps in order to get them to build?

Michaelhobo committed:
transition not erroring, but it doesn't apply correct sdk_config

Michaelhobo committed:
Get sdk_config.h transition working

Taken from https://github.com/thekyz/snippets/tree/master/cc_binary_selectable_copts, with a few minor edits

Michaelhobo committed:
remove stale comments

Michaelhobo commented:
I got it working, thanks for taking a look!
